Make autobump pr creation safer

dcsobral opened the pull request:
The PR creation task is now responsible for handling existing PRs check
and PR updates. Pull request and commit messages were improved, as well
as error messages.

These changes required extending github4s to add draft PR support, and
implement a number of methods that were not available. It also added
support to github's v4 API, which is graphql-based.

dcsobral committed:
Make autobump pr creation safer

The PR creation task is now responsible for handling existing PRs check
and PR updates. Pull request and commit messages were improved, as well
as error messages.

These changes required extending github4s to add draft PR support, and
implement a number of methods that were not available. It also added
support to github's v4 API, which is graphql-based.

dcsobral commented on build.sbt:
```diff
@@ -6,12 +6,14 @@ ThisBuild / scalaVersion := "2.12.10"
     name := "sbt-precog-core",
-    scalacStrictMode := false)
+    scalacStrictMode := false,
+    /*Compile/sourceGenerators += graphqlCodegen.value*/)
```
Remove this.

dcsobral commented on .gitignore:
```diff
@@ -6,6 +6,8 @@ tags
 
 /core/src/main/resources/core/common-secrets.yml
 
+.graphqlconfig
```
This is a configuration file for an IntelliJ plugin/node app. It contains the authorization token, which is why I'm forcibly ignoring it.

dcsobral commented on core/build.sbt:
```diff
@@ -3,8 +3,17 @@ libraryDependencies += "org.yaml" % "snakeyaml" % "1.26"
 addSbtPlugin("com.codecommit"    % "sbt-github-actions" % "0.1-58d9629")
 addSbtPlugin("de.heikoseeberger" % "sbt-header"         % "5.4.0")
-addSbtPlugin("com.dcsobral"      % "sbt-trickle"        % "0.2.4")
+addSbtPlugin("com.dcsobral"      % "sbt-trickle"        % "0.3-37c1a2c")
```
I'm waiting to be certain of what the API needs to look like before I make a release.

dcsobral commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      pullRequest <- EitherT(createOrUpdatePullRequest(branchName, changes, label, maybePullRequest))
+    } yield pullRequest
+
+    app.leftSemiflatMap(_.warn(log)).value.unsafeRunSync().isRight
```
I'm considering returning `IO` here, to make it testable. Thoughts?

djspiewak replied:
In general, you want to push `IO` as far out to the leaves as possible. In this case, that would mean returning `IO` (i.e. *not* using `unsafe` anything) all the way to the setting/task definitions, where you would of course need to use the `unsafe` function to run it. There isn't *that* much of a difference in terms of testability, but it does help a lot in terms of reasoning when and where effects are run.

So tldr, I do think returning `IO` here would be better.

dcsobral commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      } yield updateResult.bimap(_ => UpdateError, _ => (dirFile, branchName, oldestPullRequest, changes, label))
+    }
+
+    // TODO: add changes to commit message?
```
@djspiewak Feedback on this, please.

dcsobral commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      _ <- IO(log.info(s"Removed branch ${pullRequest.base.map(_.ref)} from $owner/$repoSlug"))
+    } yield NotOldest(oldest, pullRequest).asLeft[PullRequestDraft]
+
+    // TODO: Update pull request description?
```
@djspiewak Feedback on this, please. I'm wondering where I'm going to do it -- at the point I made a decision on create/update, it is too soon since I don't know if `push` will succeed.

djspiewak replied:
If the push *doesn't* succeed, then you just drop it. If we were to make changes to the existing PR, then we would probably do them *after* the push.

I think it's fine to not do anything to the existing PR description. We aren't putting information into it right now anyway, so there's nothing really to update. Eventually, when we *do* add more information to the autobump PRs, we will of course want to keep them updated, but right now neither case is true.

dcsobral replied:
Actually, we are putting information there now. Just what the revision update is, but... it's going there.

dcsobral commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      .map(_.headOption)
+  }
+
+  // TODO: use Last-Modified/If-Modified-Since
```
Remove TODO.

djspiewak commented on build.sbt:
```diff
@@ -6,12 +6,14 @@ ThisBuild / scalaVersion := "2.12.10"
 lazy val root = project
   .in(file("."))
   .aggregate(core, artifact, plugin)
+  .settings(name := "sbt-precog-root")
```
Worth noting that we *may* have some old infrastructure which hardcodes `root` for the root project. I think I replaced all of that with `/` but I may have missed something.

dcsobral replied:
This is just for sbt-precog. I had forgotten about it, tbh. I just did it because IntelliJ is annoying -- every time it reloads the project, it replaces the project name with root's name, so I end up trying to pick which project called "root" I want to load.

dcsobral replied:
Wait... what do you mean by replacing with `/`?

djspiewak replied:
In sbt commands, `/` is an alias for the root project, whatever it may be named.

dcsobral replied:
How can it be an alias for root _and_ a separator for scopes?

djspiewak replied:
It's only an alias for root in the shell, and I *think* only in the `project` command. So you can run `project /`, but I don't think you can do something like `//test` and run `test` in root (maybe you can… dunno).

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  sealed trait Warnings {
+    def warn(log: Logger): IO[Unit]
+  }
+  case object UpdateError extends Warnings {
```
Random style note: we usually put members of an ADT under its companion object. So this would be in `object Warnings`, for example.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+import scala.util.matching.Regex
+
+object AutoBump {
+  sealed trait Warnings {
```
When encoding an ADT, it's usually best to make the parent type `extend Product with Serializable` so that your type errors don't look horrible.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      log.warn("this usually means some other repository updated the pull request before this one")
+    }
+  }
+  case class NotOldest(oldest: PullRequestDraft, draft: PullRequestDraft) extends Warnings {
```
All case classes should be `final`. Always.

dcsobral replied:
And it is rather annoying that they aren't. :/

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+
+    def unapply(arg: String): Option[ChangeLabel] = labelPattern.findFirstIn(arg).flatMap(apply)
+  }
+  case object RevisionChange extends ChangeLabel("version: revision")
```
One of the advantages to putting things into companion objects is you can simplify these names. Rather than `RevisionChange`, you'll get just `Revision`, inside of `ChangeLabel`.

dcsobral replied:
True. Curiously, nowhere in the code actually directly refers to these. I struggled on whether to make the code in `SbtPrecogBase` that decides what the label should be directly refer to these or not. I left this as is to avoid circular references, though both ADTs really ought to be moved out of `AutoBump`.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  case object FeatureChange extends ChangeLabel("version: feature")
+  case object BreakingChange extends ChangeLabel("version: breaking")
+
+  implicit class Unpack[F[_]: Monad, A](response: F[GHResponse[A]]) {
```
Minor nit: `UnpackSyntax` is probably slightly more descriptive

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  case object BreakingChange extends ChangeLabel("version: breaking")
+
+  implicit class Unpack[F[_]: Monad, A](response: F[GHResponse[A]]) {
+    def unpack: F[Either[GHException, A]] = EitherT(response).map(_.result).value
```
Why not just `response.map(_.map(_.result))`?

dcsobral replied:
I dislike nesting maps. I get lost...

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  val LinkRelation: Regex = """<(.*?)>; rel="(\w+)"""".r
+  val LabelPattern: Regex = "version: \\w+".r
+
+  def autoBumpCommitTitle(author: String) = s"Applied dependency updates by $author"
```
Minor nit: public members should have explicitly declared types.

dcsobral replied:
And IntelliJ annoys me every time I don't do it, so I don't understand why it didn't do so here. That said, this used to be `val AutoBumpCommitTitle`, in which case we _should_ leave the type out to get constant semantics. That's *got* to be the most annoying feature of Scala in my list.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+
+  def autoBumpCommitTitle(author: String) = s"Applied dependency updates by $author"
+
+  def autoPage[F[_]: Sync, T](first: Pagination)
+                             (call: Pagination => F[Either[GHException, GHResult[List[T]]]]): Stream[F, T] = {
```
Style nit! This is incorrectly formatted. Actually, most of your long signatures are.

We generally wrap signatures by wrapping all blocks to a double-indented context, *including* the return type, which we wrap before the `:`. Thus, your signature should be the following:

```suggestion
  def autoPage[F[_]: Sync, T](
      first: Pagination)(
      call: Pagination => F[Either[GHException, GHResult[List[T]]]])
      : Stream[F, T] = {
```

Basically as a rule of thumb, if you're vertically aligning *anything* other than sbt's dependency `%` operators, it's not conventional.

dcsobral replied:
Ugh, I hate dangling return types. :(

dcsobral replied:
If we had a scalastyle configs, IntelliJ would have pointed that stuff out to me.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  /** Decodes github's "Link" header into a map */
+  def getRelations(headers: Map[String, String]): Map[String, (Int, Int)] = {
+    val relations = for {
+      header <- headers.map { case (k, v) => k.toLowerCase -> v }.collect { case ("link", header) => header }
```
More random style nits… *Either* use `.` and `(`, or use <kbd>space</kbd> and `{`, but not a mixture of the two. So the above should be the following if you want to keep it one line:

```suggestion
      header <- headers map { case (k, v) => k.toLowerCase -> v } collect { case ("link", header) => header }
```

Or you can break it up into two lines for readability, but in either case you're going to end up not using `.`.

The above actually nicely illustrates *why* this rule is something you should adhere to: depending on syntactic context, `}.` doesn't do what you think it does. In this case, it *is* doing the "right" thing, but in other contexts it won't, and the rules around it are very subtle. Not quite as bad as in Ruby, but still bad. Just avoid it.

dcsobral replied:
I'm well familiar with parenthesis vs braces in Scala. I thought you didn't like operator notation on non-symbolic methods.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+    getPullRequests
+      .evalFilter(pullRequest => getLabels(pullRequest.number).map(labels => isAutoBump(pullRequest, labels)))
+      .head
+      .compile.toList
+      .map(_.headOption)
```
```suggestion
      .compile.last
```

dcsobral replied:
Last? Steam is returning a list of prs <<pr1, pr2, pr3, ..., prn>>, where pr1 is the oldest and prn is the newest, and I want the oldest, which is what the code was doing. `last` seems to do very much the opposite,no?

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+    }
+  }
+
+  // FIXME: current github4s api does not support Some(pagination)
```
What does this `FIXME` mean? Is this something wrong *here* or in github4s?

dcsobral replied:
Both. It is wrong here because github4s does not support pagination in this particular API. In fact, I filed an issue there, and it's fixed on master.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      .draftPullRequest(
+        owner,
+        repoSlug,
+        NewPullRequestData(autoBumpCommitTitle(authorRepository),
```
Style nit: put `autoBumpCommitTitle(authorRepository)` on the next line. Basically if you're wrapping actuals, either wrap all or none.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      for {
+        _ <- run(s"git add .", log, merge = true, workingDir = Some(dirFile))
+        result <- run(
+          Seq("git", "commit", "-m", autoBumpCommitTitle(authorRepository)),
```
So what we *used* to do here is build a mini changelog from the origin repo(s). Basically, grab the origin repo, do a `git log vXXX..HEAD`, filter to the merge commits descending from each other (I can't remember exactly how this works, but git has a flag for it) and take the commit messages (which are derived from pull request titles by `sdmerge`).

I'd like to revive this at some point, but it's harder with trickle since you'd need to go clone *those* repos, and also the changelog thing wasn't really perfect anyway. This is fine for now, and we can refine it over time.

dcsobral replied:
It's even harder because, the way it was done, we knew for sure what the previous version was. It's trickier now, particularly with autobump PR updates. As for checking out... we don't need to do that. We can query github.

dcsobral replied:
OH, another thing... originally, every bump corresponded to a single repository version bump. Now we are combining bumps from multiple repositories.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+    } yield NotOldest(oldest, pullRequest).asLeft[PullRequestDraft]
+
+    // TODO: Update pull request description?
+    def createOrUpdatePullRequest(branchName: String,
+                                  changes: List[String],
+                                  changeLabel: ChangeLabel,
+                                  maybePullRequest: Option[PullRequestDraft]): IO[Either[Warnings, PullRequestDraft]] = {
```
Another style example:

```suggestion
    def createOrUpdatePullRequest(
        branchName: String,
        changes: List[String],
        changeLabel: ChangeLabel,
        maybePullRequest: Option[PullRequestDraft])
        : IO[Either[Warnings, PullRequestDraft]] = {
```

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+  object ChangeLabel {
+    val values: List[ChangeLabel] = List(RevisionChange, FeatureChange, BreakingChange)
+    val order: Map[ChangeLabel, Int] = values.zipWithIndex.toMap
+    implicit val ordering: Ordering[ChangeLabel] = Ordering.by(order)
```
Minor aside: we have Cats in scope here. Rather than using `Ordering`, it's probably better to just use `Order`. I'm not super-fussed about it either way.

dcsobral replied:
My mind was about to go into shutdown, and I'm not familiar enough with Cats yet. Hell, even dragging out the details of `Ordering` was a major ordeal at that point.

djspiewak commented on core/src/main/scala/slamdata/AutoBump.scala:
```diff
@@ -0,0 +1,369 @@
+      .as(lines.toList)
+  }
+
+  private def run_?(command: Seq[String],
```
Suffix operators on alphabetical identifiers. Now there's a style I haven't seen [in a long, long time…](https://github.com/lift/framework)

dcsobral replied:
Nothing else occurred to me, I'm afraid. It felt wrong and icky, but also understandable. I'm still a bit disgruntled that `!` and `?` cannot appear as suffixes to alphabetical identifiers, ten years later. :)

djspiewak commented on core/src/main/scala/slamdata/SbtPrecogBase.scala:
```diff
@@ -456,131 +442,58 @@ abstract class SbtPrecogBase extends AutoPlugin {
+                isBreaking ||= testBreaking
+                log.info(s"Updated ${getChange(testRevision, testBreaking)} $dependencyRepository $currentVersion -> $newRevision")
+              case None                  =>
+                // TODO: use scalafix to change build.sbt
```
That would be cool…

djspiewak commented on core/src/main/scala/slamdata/algebras/GitData.scala:
```diff
@@ -0,0 +1,21 @@
+import github4s.GithubResponses.GHResponse
+import github4s.domain.Ref
+
+trait GitData[F[_]] extends github4s.algebras.GitData[F] {
```
Trivia note! This is the very first time I've ever worked with or reviewed code that is *actually* using a finally tagless encoding.

djspiewak commented on core/src/main/scala/slamdata/domain/PullRequestDraft.scala:
```diff
@@ -0,0 +1,89 @@
+
+import github4s.domain.{PullRequest, PullRequestBase, User}
+
+case class PullRequestDraft(id: Int,
```
Should be `final`. Also class formals should be formatted analogously to method formals: when split across lines, double-indent rather than vertically aligning, with `extends` and `with` being similarly double-indented and on their own line(s).

djspiewak commented on core/src/main/scala/slamdata/domain/PullRequestDraft.scala:
```diff
@@ -0,0 +1,89 @@
+                            head: Option[PullRequestBase],
+                            user: Option[User],
+                            assignee: Option[User]
+                           )
```
Another random style nit: we usually don't favor dangling `)` in this fashion. Instead, we put it at the end of the previous line. This distinguishes the scope a bit more from `}` (which we *do* push to the next line!), where the newline inference semantics are different.

djspiewak commented on core/src/main/scala/slamdata/domain/PullRequestDraft.scala:
```diff
@@ -0,0 +1,89 @@
+                           )
+
+object PullRequestDraft {
+//  implicit def toPullRequest(pr: PullRequestDraft): PullRequest = PullRequest(
```
A rare valid use of implicit coercion, btw.

djspiewak commented on core/src/main/scala/slamdata/domain/PullRequestDraft.scala:
```diff
@@ -0,0 +1,89 @@
+  def draft: Boolean
+}
+
+case class DraftPullRequestData(title: String,
```
Moar final…

djspiewak commented on core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala:
```diff
@@ -0,0 +1,86 @@
+import github4s.http.HttpClient
+import precog.algebras.GitData
+
+class GitDataInterpreter [F[_] : Sync](impl: github4s.algebras.GitData[F])
+                                      (implicit client: HttpClient[F], accessToken: Option[String])
+    extends GitData[F] {
```
```suggestion
class GitDataInterpreter[F[_]: Sync](
    impl: github4s.algebras.GitData[F])(
    implicit client: HttpClient[F],
    accessToken: Option[String])
    extends GitData[F] {
```

Why is the access token `implicit`? Also the client? But really especially the access token.

dcsobral replied:
I've no idea. I just copy & pasted the declaration from github4s, taking it on faith that it made sense.

dcsobral replied:
I guess just to reduce duplication initializing all these classes on `Github`. <sigh>

djspiewak commented on core/src/main/scala/slamdata/interpreters/PullRequestsInterpreter.scala:
```diff
@@ -0,0 +1,155 @@
+  import github4s.Decoders._
+
+  /** Representation of the json data github graphql queries expect. */
+  case class GithubQuery[A <: GraphQLQuery](query: A, variables: A#Variables)
```
First off, `final`. Second, this is a surprisingly tricky type particularly taken into context with the implicits in the companion object. Is it actually strictly necessary?

djspiewak reviewed (commented):
Mostly a ton of style nits and Scala best practice points. Nothing too major. I would be fine merging this as-is, in fact. Nice work.

Regarding your major points on PR descriptions, I definitely think we should add more detail to them down the line, but right now we should just get this out the door and refine it later. It would be really nice to be able to see at a glance (when reviewing trickle PRs) what is going on, similar to what we had before with autobumps, but it's workable without it.

dcsobral commented:
> Regarding your major points on PR descriptions, I definitely think we should add more detail to them down the line, but right now we should just get this out the door and refine it later. It would be really nice to be able to see at a glance (when reviewing trickle PRs) what is going on, similar to what we had before with autobumps, but it's workable without it.

We do tag every release, don't we? I just realized getting the commit logs is easier than I thought. I was too focused on how the old autobump worked to realize I can simply use github4s to retrieve the logs without checking out the repositories.

djspiewak commented:
> We do tag every release, don't we? I just realized getting the commit logs is easier than I thought. I was too focused on how the old autobump worked to realize I can simply use github4s to retrieve the logs without checking out the repositories.

Ah that's a good point. Yes, we do, barring bugs. Sometimes weird things happen and the tags don't get created and we shouldn't fail if that's the case.

If it's easy to compute changelogs, go for it. :-) I'm also fine just doing that in a follow-up PR.

dcsobral committed:
Pull request feedback

djspiewak commented:
Needs merge from master btw

dcsobral committed:
Nitpick stuff

dcsobral committed:
Extract runner and make createPullRequest return IO

All parameters to createPullRequest have been moved to the class
constructor, so that all nested methods could be moved out of
createPullRequest (without adding the repeated parameters), so
that each maybe tested individually.

All run methods were moved into a Runner class. Fixed an issue
detected by tests for that class (mutable "Buffer" was being
initialized outside IO).

dcsobral committed:
Add tests for ChangeLabel

djspiewak commented on core/src/main/scala/slamdata/Github.scala:
```diff
@@ -0,0 +1,63 @@
+  val gists: Gists[F]                    = new GistsInterpreter[F]
+  val issues: Issues[F]                  = new IssuesInterpreter[F]
+  val activities: Activities[F]          = new ActivitiesInterpreter[F]
+  val gitData: GitData[F]                = new GitDataInterpreter[F](
```
IntelliJ's auto-formatter kinda went nuts here…

dcsobral replied:
Actually, this file was copied from github4s, and I just didn't bother reformatting it.

djspiewak replied:
Gotcha

djspiewak commented on core/src/test/scala/precog/RunnerSpec.scala:
```diff
@@ -0,0 +1,164 @@
+  "Fluent builders" should {
+    "be equivalent to case class build" in {
+      val runner = Runner[IO](log)
+      runner === Runner[IO](log, false, None, Map.empty, Seq.empty)
+      runner.stderrToStdout === Runner[IO](log, merge = true)
+      runner.cd(file("./target")) === Runner[IO](log, workingDir = Some(file("./target")))
+      runner.withEnv("X" -> "Y", "W" -> "Z") === Runner[IO](log, env = Map("X" -> "Y", "W" -> "Z"))
+      runner.hide("password") === Runner[IO](log, hide = Seq("password"))
```
We strongly prefer a `mustEqual` style due to the fact that `===` is a Cats equality operator.

djspiewak commented on core/src/test/scala/slamdata/AutoBumpTest.scala:
```diff
@@ -0,0 +1,45 @@
+
+import org.specs2.mutable.Specification
+
+class AutoBumpTest extends Specification {
```
`AutoBumpSpec`

dcsobral replied:
Yeah. I'm letting IntelliJ create it, and it has a weird default. Well, perhaps not _weird_, just not well suited to Specs.

djspiewak commented:
LGTM. Merging as-is. Any other changes we can address in a follow-up PR if necessary
